### PR TITLE
feat(async-csv): setup CSV export logic in celery

### PR DESCRIPTION
**Pull request by leeandher**
## Update

This PR is ready for re-review, it now uses the `tempfile` module for it's cleanup, and saves the `File` object to the database, updating `ExportedData` in the process (date_completed, date_expired, file).

---

This PR contains the abstracted code to create CSVs on the backend by using existing routes in the API.

**How it works**
- POST endpoint (`organizations/org/data-export/`) triggers a celery task after creating a record in the ExportData table.
- Celery task hits the appropriate endpoint (so far only <s>**Discover v1**</s> **Issues by Tags**, and receives the raw payload)
- Helper functions convert the data into a CSV

**To do in this PR**
- [x] Upload files to filestore backend solution
- [x] Update the entries in the ExportData table

---

**API-692**

### CODE REVIEW

File: src/sentry/api/serializers/models/exporteddata.py
Comment (leeandher) [inline src/sentry/api/serializers/models/exporteddata.py]: @wedamija suggested this type of implementation in the previous PR, to reduce querying

File: src/sentry/tasks/data_export.py
Comment (leeandher) [inline src/sentry/tasks/data_export.py]: This was obviously just for local testing, and will be removed/edited before merging

File: src/sentry/tasks/data_export.py
Comment (leeandher) [inline src/sentry/tasks/data_export.py]: This is the best way I could come up with to abstract this 'Export to CSV' task.

 Each use case will have different processing logic and by massaging the payloads, we can use the same `convert_to_csv` function (and eventually, `upload_to_filestore` function) for all of them.

**Comment by leeandher** [comment]
Right now the API writes to disk, but the routes are disabled by default so they can't be accessed in production/staging

The **next** step is to figure out file storage (upload/download) situation. If anyone has any tips on using the `File` model with GCS that'd be great :s

File: src/sentry/tasks/data_export.py
Comment (scefali) [inline src/sentry/tasks/data_export.py]: @leeandher is it really an endless loop though?

File: src/sentry/tasks/data_export.py
Comment (scefali) [inline src/sentry/tasks/data_export.py]: @leeandher is importing here necessary?

File: src/sentry/tagstore/snuba/backend.py
Comment (markstory) [inline src/sentry/tagstore/snuba/backend.py]: What is this for?

File: src/sentry/tasks/data_export.py
Comment (markstory) [reply src/sentry/tasks/data_export.py]: If it has no end you'll never be able to send a CSV file. :smile:

File: src/sentry/tasks/data_export.py
Comment (markstory) [inline src/sentry/tasks/data_export.py]:
```suggestion
    file_details = u"{}-{}".format(payload["project_slug"], key)
```
Just incase the data has unicode in it.

File: src/sentry/tasks/data_export.py
Comment (markstory) [inline src/sentry/tasks/data_export.py]: Typo in results.

File: src/sentry/tasks/data_export.py
Comment (scefali) [reply src/sentry/tasks/data_export.py]: @leeandher I think we should put a timeout on this iteration so we don't do it endlessly :)

File: src/sentry/tasks/data_export.py
Comment (leeandher) [reply src/sentry/tasks/data_export.py]: I put the imports specific to whatever process function is running (i.e. When i get to Discover Processing, it'll have different imports specific to Discover.)

File: src/sentry/tagstore/snuba/backend.py
Comment (leeandher) [reply src/sentry/tagstore/snuba/backend.py]: This allows me to pass in `offset` without breaking wherever it's already implemented (otherwise I'd have to add `offset=0` to wherever it's used)

Comment (leeandher) [reply src/sentry/tagstore/snuba/backend.py]: Or add `offset=0` as the default param, which didn't seem right to me

File: src/sentry/tasks/data_export.py
Comment (leeandher) [reply src/sentry/tasks/data_export.py]: Endlessly was a really dumb word to use, I just mean it's a `while True` loop with a `break` statement

File: src/sentry/tasks/data_export.py
Comment (scefali) [reply src/sentry/tasks/data_export.py]: @leeandher In general, we should put imports at the top of the file. Sometimes this causes errors which is why we do it in a function.

**Comment by leeandher** [comment]
This code requires https://github.com/getsentry/sentry/pull/16912 to be merged first, it has the same migration commit

File: src/sentry/tasks/data_export.py
Comment (leeandher) [reply src/sentry/tasks/data_export.py]: alright, will do

File: src/sentry/tasks/data_export.py
Comment (wedamija) [inline src/sentry/tasks/data_export.py]: I don't think we really need this. Why not just use `tempfile.TemporaryFile`?

File: src/sentry/tasks/data_export.py
Comment (wedamija) [inline src/sentry/tasks/data_export.py]: We could just use `tempfile.NamedTemporaryFile` here, and only ever open the file once. Then we can also use it as a context manager, and so when we're done it'll be cleaned up automatically. Same for the dict writer, we should probably just instantiate it out here, write the header and just keep writing to it. That way we avoid passing the `include_header` param too.

File: src/sentry/tasks/data_export.py
Comment (leeandher) [reply src/sentry/tasks/data_export.py]: It's limited by snuba, so a list of 1000 dicts

File: src/sentry/tasks/data_export.py
Comment (leeandher) [reply src/sentry/tasks/data_export.py]: That's actually exactly what I thought too when I saw `NamedTemporaryFile` in the docs, but I was trying to follow what was done in `src/sentry/tasks/assemble.py` 

I'll refactor it!

File: src/sentry/tagstore/snuba/backend.py
Comment (markstory) [reply src/sentry/tagstore/snuba/backend.py]: The explicit parameter enables better mistake detection as you won't get schema validation errors from snuba farther down the callstack. Spreading kwargs into snuba has become a problematic pattern over time as it has made tracking back what parameters each call is using much harder to track down impeding refactoring.

You should also update the interfaces defined in `tagstore/base.py` as well.

Comment (leeandher) [reply src/sentry/tagstore/snuba/backend.py]: @markstory Addressed in one of the latest commits, I wasn't sure if it should be added as a default for the definition of `get_group_tag_value_iter()`, or added as offset=0 to every time it's used

I chose the latter but let me know if I should do the opposite.

File: src/sentry/tasks/data_export.py
Comment (leeandher) [inline src/sentry/tasks/data_export.py]: @wedamija I'm trying to abstract as much as possible to be used in the other `processing` functions (e.g. Billing, Discover v2), let me know if this is a good way to do that.

File: src/sentry/tasks/data_export.py
Comment (leeandher) [inline src/sentry/tasks/data_export.py]: Probably unnecessary but I like how it looks, lmk if it should get scrapped

**Review by scefali — APPROVED** [review]
LGTM but see if you can get some feedback from Dan and test it on stage

**Review by wedamija — COMMENTED** [review]
We need some tests around the various functions in `data_export.py`

File: src/sentry/tagstore/base.py
Comment (markstory) [inline src/sentry/tagstore/base.py]: Is there any usage of callbacks parameter? Adding the parameter here could break existing callers.

Comment (leeandher) [reply src/sentry/tagstore/base.py]: I've added the `offset=0` to everywhere this function is used, and the `callbacks` parameter is always named or ignored.

Comment (markstory) [reply src/sentry/tagstore/base.py]: Making this parameter required complects all the callsites. If this was an optional parameter with a default value you would have been able to do fewer changes. Given that you've done the work already there is no point in going back, but might be something to keep in mind for next time.

File: tests/sentry/tasks/test_data_export.py
Comment (markstory) [inline tests/sentry/tasks/test_data_export.py]: Why do you need to call setup here?

File: tests/sentry/tasks/test_data_export.py
Comment (markstory) [inline tests/sentry/tasks/test_data_export.py]: I think you want setUp(). That is the method unittest.TestCase uses. Remember to call super() as well.

Comment (leeandher) [reply tests/sentry/tasks/test_data_export.py]: Gonna be honest, didn't even know that was a built-in method -- I was doing this because I though that's how I could separate things in tests

I'll refactor it!

File: src/sentry/tagstore/base.py
Comment (leeandher) [reply src/sentry/tagstore/base.py]: I think you're right regardless, I just went back and reverted it. It makes more sense to leave the default.

**Pull request merged**
